sycl/test/check_device_code/vector/vector_bf16_builtins.cpp

```diff
@@ -14,361 +20,259 @@ using namespace sycl::ext::oneapi;
-// CHECK-NEXT:    [[REF_TMP_SROA_0_0_I:%.*]] = phi i16 [ [[TMP7]], [[IF_END18_I_I]] ], [ [[TMP2]], [[IF_END6_I_I]] ], [ -32768, [[IF_END10_I_I]] ], [ [[SPEC_SELECT_I]], [[LAND_LHS_TRUE_I_I]] ]
-// CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 2, ptr nonnull [[AGG_TMP10_I]]), !noalias [[META12]]
-// CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 2, ptr nonnull [[AGG_TMP111_I]]), !noalias [[META12]]
-// CHECK-NEXT:    [[CALL5_I:%.*]] = call spir_func noundef align 2 dereferenceable(2) ptr addrspace(4) @_ZN4sycl3_V13vecINS0_3ext6oneapi8bfloat16ELi2EEixIS4_EENSt9enable_ifIXsr3stdE9is_same_vIT_S4_EERS4_E4typeEi(ptr addrspace(4) noundef align 4 dereferenceable_or_null(4) [[AGG_RESULT]], i32 noundef [[CONV_I]]) #[[ATTR7]]
-// CHECK-NEXT:    store i16 [[REF_TMP_SROA_0_0_I]], ptr addrspace(4) [[CALL5_I]], align 2, !tbaa [[TBAA10]]
-// CHECK-NEXT:    [[INC_I]] = add nuw nsw i64 [[I_0_I]], 1
-// CHECK-NEXT:    br label [[FOR_COND_I]], !llvm.loop [[LOOP19:![0-9]+]]
-// CHECK:       _ZN4sycl3_V13ext6oneapi12experimental4fminINS0_3vecINS2_8bfloat16ELi2EEES7_Li2ELi2EEENSt9enable_ifIXaaaa24is_vec_or_swizzle_bf16_vIT_E24is_vec_or_swizzle_bf16_vIT0_EeqT1_T2_ENS5_IS6_XT1_EEEE4typeES9_SA_.exit:
-// CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 4, ptr nonnull [[AGG_TMP2]])
-// CHECK-NEXT:    call void @llvm.lifetime.end.p0(i64 4, ptr nonnull [[AGG_TMP13]])
+// CHECK-NEXT:    [[VEC_ADDR_I_I_I_I8_I:%.*]] = alloca <2 x float>, align 8
+// CHECK-NEXT:    [[DST_I_I_I_I9_I:%.*]] = alloca [2 x %"class.sycl::_V1::ext::oneapi::bfloat16"], align 2
+// CHECK-NEXT:    [[VEC_ADDR_I_I_I_I2_I:%.*]] = alloca <2 x i16>, align 4
+// CHECK-NEXT:    [[DST_I_I_I_I3_I:%.*]] = alloca [2 x float], align 4
+// CHECK-NEXT:    [[VEC_ADDR_I_I_I_I_I:%.*]] = alloca <2 x i16>, align 4
+// CHECK-NEXT:    [[DST_I_I_I_I_I:%.*]] = alloca [2 x float], align 4
+// CHECK-NEXT:    [[TMP0:%.*]] = load <2 x i16>, ptr [[A]], align 4
+// CHECK-NEXT:    [[TMP1:%.*]] = load <2 x i16>, ptr [[B]], align 4
+// CHECK-NEXT:    tail call void @llvm.experimental.noalias.scope.decl(metadata [[META8:![0-9]+]])
+// CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 4, ptr nonnull [[VEC_ADDR_I_I_I_I_I]]), !noalias [[META11:![0-9]+]]
+// CHECK-NEXT:    call void @llvm.lifetime.start.p0(i64 8, ptr nonnull [[DST_I_I_I_I_I]]), !noalias [[META11]]
+// CHECK-NEXT:    [[VEC_ADDR_ASCAST_I_I_I_I_I:%.*]] = addrspacecast ptr [[VEC_ADDR_I_I_I_I_I]] to ptr addrspace(4)
+// CHECK-NEXT:    [[DST_ASCAST_I_I_I_I_I:%.*]] = addrspacecast ptr [[DST_I_I_I_I_I]] to ptr addrspace(4)
+// CHECK-NEXT:    store <2 x i16> [[TMP0]], ptr [[VEC_ADDR_I_I_I_I_I]], align 4, !tbaa [[TBAA14:![0-9]+]], !noalias [[META11]]
+// CHECK-NEXT:    call spir_func void @__devicelib_ConvertBF16ToFINTELVec2(ptr addrspace(4) noundef [[VEC_ADDR_ASCAST_I_I_I_I_I]], ptr addrspace(4) noundef [[DST_ASCAST_I_I_I_I_I]]) #[[ATTR6:[0-9]+]], !noalias [[META11]]
```

Comment:
Why new is better? Do you have any performance measurements? I see we're doing extra converts and also I'm not sure this "optimization" doesn't change the resulting value vs native bfloat16 compute.

Reply:
> Why new is better?

Previously, on device, we were doing element-by-element computation in math builitins (for `vec<bfloat16>`). Now, after this change, we convert BF16 `sycl::vec` to `sycl::vec<float>`, perform computations on `sycl::vec<float>` using spirv builitins (which should vectorize operations underneath), and convert the result back to `vec<bfloat16>`. On device, the conversion between `vec<float>` and `vec<bfloat16>` happens using OpConvertBF16ToFINTEL and OpConvertFToBF16INTEL which will also result in vectorized conversions.
Note that, currently, for BF16 scalar support in math builtins, we anyway convert back and forth to float: https://github.com/uditagarwal97/llvm/blob/sycl/sycl/include/sycl/ext/oneapi/experimental/bfloat16_math.hpp#L342

> Do you have any performance measurements?

I don't think it's necessary - generated device code is clear enough.

> I'm not sure this "optimization" doesn't change the resulting value vs native bfloat16 compute

I'm not exactly sure what you meant by this. Looking at the device code, the return value of `__spirv_ocl_fmin` is converted to `vec<bfloat16>` and returned, which seems correct to me.

Reply:
We vectorize over entire sub-group, not across vector elements. I doubt the improvement in the operation itself would be as significant as you expect it to be and possibly not big enough to justify double convert (bf16->float->bf16).

Reply:
Regarding overhead of convert, we do the conversion from `float` <--> `bfloat` even before this optimization. In fact, earlier we used to do element-by-element conversion but now we use OpConvertBF16ToFINTEL etc. for vectorized conversions, which IIUC, should strictly be faster.